h/services/annotation_delete.py

```diff
@@ -37,14 +39,28 @@ def delete_annotations(self, annotations):
 
     def bulk_delete(self):
         """Expunge annotations marked as deleted from the database."""
-        self.request.db.query(Annotation).filter_by(deleted=True).filter(
-            # Deletes all annotations flagged as deleted more than 10 minutes ago. This
-            # buffer period should ensure that this task doesn't delete annotations
-            # deleted just before the task runs, which haven't yet been processed by the
-            # streamer.
-            Annotation.updated
-            < datetime.utcnow() - timedelta(minutes=10)
-        ).delete()
+        self.request.db.execute(
+            delete(Annotation).where(
+                Annotation.id.in_(
+                    select(
+                        select(Annotation.id)
+                        .where(Annotation.deleted.is_(True))
+                        # Wait ten minutes before expunging an annotation to
+                        # give the streamer time to process the deletion.
+                        .where(
+                            Annotation.updated
+                            < datetime.utcnow() - timedelta(minutes=10)
+                        )
+                        # Only expunge up to 1000 annotations at a time to
+                        # avoid long-running DB queries. This method is called
+                        # periodically so eventually all deleted annotations
+                        # will get expunged.
+                        .limit(1000)
+                        .cte()
```

Comment:
Not sure if this CTE here is needed, we are already using a subquery

Reply:
It's actually needed, you can't just pass a `select()` as the argument to `select()`, the `cte()` transforms it into an acceptable type